File: day-06/main.hs
import Debug.Trace
import Control.Monad
import System.IO

--countWaysToBeat' e r = length $ filter (\t -> t * (e - t) > r) [0..e]
countWaysToBeat :: Int -> Int -> Int
countWaysToBeat e r = ceiling ((e' + q)/2) - floor ((e' - q)/2) - 1
  where
    e' = fromIntegral e
    q = sqrt . fromIntegral $ e*e - 4*r

aoc path = do
  print path
  numbers :: [[Int]] <- map (map read . drop 1 . words) . lines <$> readFile path
  guard $ length numbers == 2
  let [times, distances] = numbers
  let races = zip times distances
  let ways = flip map races $ uncurry countWaysToBeat
  print $ product ways
  let [e, r] = map (read . join . map show) numbers
  print $ countWaysToBeat e r

main = do
  forM_ ["../inputs/sample-day-06.txt", "../inputs/input-day-06.txt"] $ aoc

File: day-06/main.py
import os

def run(input_file):
    with open(input_file) as f:
        lines = [s.split(":")[1] for s in f.readlines()]
        races = zip(*(map(int, s.split()) for s in lines))
        p = 1
        for (e, r) in races:
            c = 0
            for t in range(0, e):
                d = t * (e - t)
                if d > r:
                    c += 1
            p *= c
        print(p)

        c = 0
        (e, r) = (int(s.replace(" ", "")) for s in lines)
        for t in range(0, e):
            d = t * (e - t)
            if d > r:
                c += 1
        print(c)

today = os.path.dirname(os.path.realpath(__file__)).rpartition('/')[2]
run(f"../inputs/sample-{today}.txt")
run(f"../inputs/input-{today}.txt")

#d = t * (e - t)
#d > r
#r < t * (e - t)

